lib/IDE/TypeCheckCompletionCallback.cpp

```diff
@@ -92,18 +93,93 @@ Type swift::ide::getTypeForCompletion(const constraints::Solution &S, Expr *E) {
       return type;
     }));
   } else {
-    Result = S.getResolvedType(E);
+    Result = S.getResolvedType(Node);
   }
 
   if (!Result || Result->is<UnresolvedType>()) {
-    Result = CS.getContextualType(E, /*forConstraint=*/false);
+    Result = CS.getContextualType(Node, /*forConstraint=*/false);
   }
   if (Result && Result->is<UnresolvedType>()) {
     Result = Type();
   }
   return Result;
 }
 
+/// If the code completion variable occurs in a pattern matching position, we
+/// have an AST that looks like this.
+/// \code
+/// (binary_expr implicit type='$T3'
+///   (overloaded_decl_ref_expr function_ref=compound decls=[
+///     Swift.(file).~=,
+///     Swift.(file).Optional extension.~=])
+///   (argument_list implicit
+///     (argument
+///       (code_completion_expr implicit type='$T1'))
+///     (argument
+///       (declref_expr implicit decl=swift_ide_test.(file).foo(x:).$match))))
+/// \endcode
+/// If the code completion expression occurs in such an AST, return the
+/// declaration of the \c $match variable, otherwise return \c nullptr.
+static VarDecl *getMatchVarIfInPatternMatch(Expr *E, ConstraintSystem &CS) {
```

Comment:
Good news is that `ConstraintSystem` already has a way to check this `ConstraintSystem::isPatternMatchingOperator`

Reply:
Nice. Thanks for catching this. Surely removes some ugly code.